### ISSUE TITLE
Some API objects should be #[repr(C)] and returned by value for stack allocation

### ISSUE THREAD

**Issue by luketpeterson** [trueagi-io/hyperon-experimental#350]
The C API currently pays a malloc penalty for just about every single object.  This is because opaque Rust types are boxed, and that box is converted to a pointer to return from the C function.

However, this imposes a performance cost on the non-Rust clients.  Eventually we could make some of the fundamental Rust types be #[repr(C)].  But in the meantime, a pointer to a struct is no different from a struct containing a pointer.

So, I propose returning structs directly instead of pointers.  For example:
```C
typedef struct atom_t {
    struct rust_atom *atom;
} atom_t;

struct atom_t atom_sym(const char *name);
```

**Comment by vsbogd** [comment]
I didn't do it from the very beginning because I wanted to be able developing any Rust API I would like to and I believed that there are Rust data types which cannot be mapped to the C one's automatically. @luketpeterson do you know if it is a case?

**Comment by vsbogd** [comment]
Also I am not sure about how huge this performance cost is. For example for the Python clients we need to copy the data into a heap anyway I believe. We probably need performance tests to measure performance issues.

**Comment by luketpeterson** [comment]
> I didn't do it from the very beginning because I wanted to be able developing any Rust API I would like to and I believed that there are Rust data types which cannot be mapped to the C one's automatically. @luketpeterson do you know if it is a case?

Certainly some Rust types aren't internally `#[repr(C)]`, but for those the answer is to embed the pointer inside the structure instead.  So a struct containing an opaque pointer, rather than a pointer to an opaque structure.

> Also I am not sure about how huge this performance cost is. For example for the Python clients we need to copy the data into a heap anyway I believe. We probably need performance tests to measure performance issues.

For the changes I propose at first, there would be absolutely no performance impact nor benefit.  It would be fundamentally the same operations.

My reasoning for wanting to make this change is as a way to future-proof the API.  ie. if/when we want to optimize the C bindings, we'd have the flexibility to do it transparently, rather than force all the clients to change their use case.  So getting the API right before Alpha means we'd have the flexibility to improve things behind the scenes.

**Comment by luketpeterson** [comment]
Looking to the future, the two objects that can be somewhat ephemeral in the API are `Atom` and `BindingsSet`.  Those are the ones that would have be biggest benefit in the longer term.

You're absolutely right that these costs are currently dwarfed by internal and algorithmic costs (working on those too).  But I'm just trying to get the API into a state where we don't have to force the clients to change their code more than necessary after Alpha.

**Comment by vsbogd** [comment]
>  So a struct containing an opaque pointer, rather than a pointer to an opaque structure.

Could you please provide a small example of code, because I don't fully understand what you mean.

> But I'm just trying to get the API into a state where we don't have to force the clients to change their code more than necessary after Alpha.

Here my experience tells me the opposite :-) Anyway there will be changes which will require clients to adapt their code to the new API. And making Rust structures to be presentable in C will not make life easier because after that we will need to keep compatibility in a Rust level (I have feeling that we will finish with having separate types for C anyway). And also we will have an additional requirements to have structures in Rust to be presentable in C. Thus I would propose stabilize Rust API first and after that experiment with making Rust API `#[repr(C)]`.

**Comment by luketpeterson** [comment]
This is what I have in mind.  e765c55f87366f0d0f362a8a578836f4c79a33f7

I think the two most important objects, from a performance standpoint, are atom_t and bindings_set_t, but IMO it makes sense to do this with all of them for consistency.

**Comment by luketpeterson** [comment]
> Thanks @luketpeterson, but replacing `*mut bindings_set_t` by `bindings_set_t` which wraps the pointer doesn't change the performance. Do you mean it is a transition change before replacing pointer inside `bindings_set_t` by `BindingsSet` instance?

Yes exactly.  This means that, in the future, we can move more meaningful contents into the bindings_set_t and the client's code can just recompile.

**Comment by luketpeterson** [comment]
This has turned into a large rabbit hole, and this experience stands as a substantial data point in favor of dumping cbindgen in favor of saferffi, since the blocking issues are due to cbindgen.

I will briefly go over the issues in the order I encountered them.

1.) If `atom_t` is a statically-sized object owned by the C caller, somehow functions like `vec_atom_get` need to be able to return a reference to an existing Atom, not necessarily a copy.  So I thought returning `*const atom_t` would be nice, but that means a pointer to a Rust `Atom` needs to be 1-to-1 translatable with a ptr to an `atom_t`, which is only possible if they have the same memory layout.

2.) Since I wanted to make `Atom` into a `#[repr(C)]` struct anyway, I figured I'd bite the bullet and just do that.  But that turned into a deep rabbit hole because a number of Rust primitives aren't `#[repr(C)]` (such as `String`, `Vec`, etc.)  I could roll my own versions, but it would take a bit of time, and `saferffi` already provides this.

3.) So, I decided to take a different tack, and I wrote this: https://crates.io/crates/repr_c_wrapper in order to provide a nice convenient encapsulation for the Rust types that could be held by C code.

4.) Unfortunately `repr_c_wrapper` relies on a macro to create the corresponding opaque C type for each Rust type.  And macros aren't parsed by default by cbindgen.  `cbindged` has a facility to expand macros - controlled with the `[parse.expand]` section of the cbindgen.toml, buuuut... It invokes the rustc macro expand feature which requires the Rust nightly tool chain.  (not available under stable tool chain)  I filed this: https://github.com/mozilla/cbindgen/issues/855 but the bug has persisted for years, so I doubt it'll be fixed any time soon.

5.) Even if the macro expansion in cbindgen *is* possible for us, I'll need to write a proc macro (another crate) in order to support pre-computing the compile-time constants needed for the C header, because the parsing done by cbindgen is really limited.

6.) In the end, I made `atom_t` capable of representing both a pointer to another Atom (borrowed) or a raw Box ptr (owned), but I feel like this is a stop-gap solution.
